File: answers/14.density_estimation/kernel_density.py
"""
Kernel Density Estimation
Non-parametric density estimation using kernels
"""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import gaussian_kde
from sklearn.neighbors import KernelDensity

print("=== Kernel Density Estimation ===\n")

# Generate data
np.random.seed(42)
data = np.concatenate([np.random.normal(0, 1, 100), np.random.normal(5, 1.5, 100)])

# KDE using scipy
kde_scipy = gaussian_kde(data, bw_method='scott')
x_range = np.linspace(-5, 10, 500)
density_scipy = kde_scipy(x_range)

# KDE using sklearn with different bandwidths
bandwidths = [0.5, 1.0, 2.0]
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

axes[0].hist(data, bins=30, density=True, alpha=0.5, label='Data')
axes[0].plot(x_range, density_scipy, 'r-', linewidth=2, label='KDE (scipy)')
axes[0].set_title('Kernel Density Estimation')
axes[0].legend()

for bw in bandwidths:
    kde = KernelDensity(bandwidth=bw, kernel='gaussian')
    kde.fit(data.reshape(-1, 1))
    log_dens = kde.score_samples(x_range.reshape(-1, 1))
    axes[1].plot(x_range, np.exp(log_dens), label=f'h={bw}')

axes[1].set_title('Effect of Bandwidth')
axes[1].legend()
plt.tight_layout()
plt.savefig('/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/14.density_estimation/kernel_density.png', dpi=150)
print("Visualization saved")
plt.close()

File: answers/14.density_estimation/kernel_density.jl
# Kernel Density Estimation in Julia
using Distributions, Random, Plots, KernelDensity

println("=== Kernel Density Estimation ===\n")
Random.seed!(42)

# Generate data (bimodal)
data = vcat(randn(100), randn(100) .+ 5)

# KDE with different bandwidths
kde_default = kde(data)
kde_small = kde(data, bandwidth=0.5)
kde_large = kde(data, bandwidth=2.0)

# Visualization
p1 = histogram(data, bins=30, normalize=:pdf, alpha=0.5, label="Data", legend=:topright)
plot!(p1, kde_default.x, kde_default.density, linewidth=2, label="KDE", color=:red)
title!(p1, "Kernel Density Estimation")

p2 = plot(kde_small.x, kde_small.density, linewidth=2, label="h=0.5", color=:blue)
plot!(p2, kde_default.x, kde_default.density, linewidth=2, label="h=default", color=:red)
plot!(p2, kde_large.x, kde_large.density, linewidth=2, label="h=2.0", color=:green)
title!(p2, "Effect of Bandwidth")

plot(p1, p2, layout=(1, 2), size=(1400, 500))
savefig("/home/titan/pdfs/notes/statisticalComputingAndReporting/groupWork/answers/14.density_estimation/kernel_density_jl.png")
println("Visualization saved")
